gensim/models/word2vec.py

```diff
@@ -1392,17 +1389,27 @@ def _raw_word_count(self, job):
         """
         return sum(len(sentence) for sentence in job)
 
-    def _check_training_sanity(self, epochs=None, total_examples=None, total_words=None, **kwargs):
+    def _check_corpus_sanity(self, corpus_iterable=None, corpus_file=None, passes=1):
+        """Checks whether the corpus parameters make sense."""
+        if corpus_file is None and corpus_iterable is None:
+            raise TypeError("Either one of corpus_file or corpus_iterable value must be provided")
+        if corpus_file is not None and corpus_iterable is not None:
+            raise TypeError("Both corpus_file and corpus_iterable must not be provided at the same time")
+        if corpus_iterable is None and not os.path.isfile(corpus_file):
+            raise TypeError("Parameter corpus_file must be a valid path to a file, got %r instead" % corpus_file)
+        if corpus_iterable is not None and not isinstance(corpus_iterable, Iterable):
+            raise TypeError("The corpus_iterable must be an iterable of list, got %r instead" % corpus_iterable)
+        if corpus_iterable is not None and isinstance(corpus_iterable, GeneratorType) and passes > 1:
+            raise TypeError(
+                f"Using a generator as corpus_iterable can't support {passes} passes. Try a re-iterable sequence.")
```

Comment:
I proposed dropping py3.5 support last December in #2713; you offered enthusiastic support; all py3.5 builds/testing was turned off in January & the issue closed. And, official CPython 3.5 reached its complete EOL a couple days ago, and won't even receive critical security patches.

Reply:
I'm very much for dropping 3.5, yes (f-strings being one of the main reasons).

That hasn't happened yet though; @mpenkov WDYT?

Reply:
If either of you think gensim development code has been supporting Py3.5, you shouldn't have approved the January items (#2713 & #2715) that removed Py3.5 from the testing infrastructure. If it's not being tested, it's unlikely to be working. 

More generally, you should just make the reasonable blanket commitment, now & forevermore, that Gensim releases never support a Python that's officially EOL, to avoid constant re-discussion like this when the bulk of the Python world has moved on.

I personally would go even further, and just decide forevermore not to have new releases support any Python that will be EOL within the next 12 months. Laggard projects/organizations can always choose to use an older release (or perform backport work on their own dime). And more generally, orgs that have difficulty upgrading their deployments to stay within ~3 years of the Python feature/fix frontier have deeper problems, that should be solved by highly-paid consultants, not by open-source developers volunteering to make their own maintenance/testing/coding tasks harder.

Alternatively, Gensim could commit to the reasoned and documented policy described in Numpy's [NER29](https://numpy.org/neps/nep-0029-deprecation_policy.html) - which (coincidentally) recommended dropping Py3.5 support for releases starting in December 2019 (about when we did) and recommends dropping Py3.6 support for releases starting in November 2020.

Reply:
4.0.0 will be the first release to NOT support Python 2.7 (never mind Python 3.5).

What we do with Py3 support going forward is up to us. Not supporting EOLed Pythons makes sense to me (with docs + `setup.py` updated to match).

I don't think we're making much use of any new Python 3.x features either – the switch is more externally-enforced. Once the "no more Python 2" river has been crossed, the version makes little difference in practice IMO beyond sysops.

Reply:
Well, I concur with your decisions 9 months ago to drop Python 3.5, and find baffling the re-consideration of the decision as if the prior discussion & tangible actions, which I've linked to for reference, never even happened. Yes, `setup.py` should be updated to sync with the prior actions, too.

Reply:
I checked `setup.py` + official docs and saw Py3.5+. I simply forgot about those old CI PRs, don't read too much into it.

I hope to clean up the overall docs structure a bit, to go with 4.0.0: an explicit split into "for users" / "for contributors" (incl. this release policy, code style, docs style…); deduplicate content from wiki vs sphinx vs github README; better content on the new main page (website index).